### PR TITLE
Towards a cleaner and more maintainable internals of NonlinearSolve.jl

### LINKED ISSUE #195: Inplace Problem Specification for Methods with `size(residual) != size(u)`
Currently, for inplace problems the residual vector is determined by the size of the input `u`. This is not necessarily correct for problems that support overconstrained / underconstrained setups. For example:
https://github.com/SciML/NonlinearSolve.jl/blob/691e10a25c0b20b3349770a4d54f93b41e5235a4/src/levenberg.jl#L269-L270

Could we potentially add a kwarg like `residual_size` which could be used to initialize this?

### PR DESCRIPTION
**Pull request by avik-pal**
Fixes #195 #158 #140 #63 

* No special jacobian logic is now present in this package
  * Delegates everything to SparseDiffTools
* Uses `ADTypes` -- constructs the ADType from the old argument style!

## TODOs:

- [x] Fix Scalar
- [x] Solvers
  - [x] Raphson
  - [x] Trust Region
  - [x] Levenberg
- [x] Make tests pass again
- [x] Verify that all allocations are necessary
  - [x] Raphson
  - [x] Trust Region
  - [x] Levenberg
- [x] Compat Entries
- [x] [upstream] Add `resid_prototype` to properly construct `fu` (https://github.com/SciML/SciMLBase.jl/pull/490)
- [x] [upstream] Add a precomputed SparsityDetection type in SparseDiffTools.jl (https://github.com/JuliaDiff/SparseDiffTools.jl/pull/259)
- [x] Forward Mode for Abstract Array (was broken before as well)

**Holding Off Merge till we have some upstream benchmarks**

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
Can we setup a PDE benchmark before merging this so we know that the change is not a performance regression? It shouldn't be, but it would be good to know that the new sparse diff stuff is on par in performance with the setup we had before.

File: src/ad.jl [hunk 1]
```diff
@@ -23,22 +23,18 @@ function scalar_nlsolve_ad(prob, alg, args...; kwargs...)
     return sol, partials
```
Comment (avik-pal) [inline src/ad.jl:23]: @ChrisRackauckas, what does this code (mathematically) exactly do? I have never messed around with ForwardDiff so not really sure.

The code doesn't work if $u$ is an SVector. The previous tests had a typo and never tested that case.

**Review by ChrisRackauckas — APPROVED** [review]

File: src/ad.jl (hunk 1 above)
Comment (ChrisRackauckas) [reply src/ad.jl:23]: It's handling the derivative rule for nonlinear solve, just skipping the dual numbers in the nonlinear solve and then applying the partials at the solution point.

Comment (avik-pal) [reply src/ad.jl:23]: Done!

File: src/ad.jl [hunk 2]
```diff
@@ -1,40 +1,61 @@
...
+    f_p = scalar_nlsolve_∂f_∂p(f, uu, p)
+    f_x = scalar_nlsolve_∂f_∂u(f, uu, p)
+
+    z_arr = -inv(f_x) * f_p
 
-    f_x = ForwardDiff.derivative(Base.Fix2(f, p), uu)
     pp = prob.p
-    sumfun = let f_x′ = -f_x
-        ((fp, p),) -> (fp / f_x′) * ForwardDiff.partials(p)
+    sumfun = ((z, p),) -> [zᵢ * ForwardDiff.partials(p) for zᵢ in z]
```
Comment (ChrisRackauckas) [inline src/ad.jl:16]: This allocates?

Comment (avik-pal) [reply src/ad.jl]: Now it doesn't for scalars (both u and p). If either isn't a scalar, it will allocate

Comment (ChrisRackauckas) [reply src/ad.jl]: and static arrays?

File: src/linesearch.jl [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    eval_f(u) = evaluate_f(f, u, p, IIP; fu)
+
+    ls.method isa Static && return LineSearchCache(eval_f, nothing, nothing, nothing,
+        convert(eltype(u), ls.α), ls)
+
+    g₀ = _mutable_zero(u)
+
+    autodiff = if iip && (ls.autodiff isa AutoZygote || ls.autodiff isa AutoSparseZygote)
+        @warn "Attempting to use Zygote.jl for linesearch on an in-place problem. Falling back to finite differencing."
+        AutoFiniteDiff()
```
Comment (ChrisRackauckas) [inline src/linesearch.jl:96]: why not fallback to ForwardDiff if it's being used elsewhere?

Comment (avik-pal) [reply src/linesearch.jl:96]: I wanted to use `VecJac`, which doesn't have the forwarddiff yet (I think?). See https://github.com/JuliaDiff/SparseDiffTools.jl/blob/04810f5caea10633504ad44071f1a6b82cb9bb5a/src/differentiation/vecjac_products.jl#L51-L52

I can fix it upstream to just construct jacobian and then multiply, and then change the default here later

**Comment by avik-pal** [comment]
Needs https://github.com/SciML/DiffEqBase.jl/pull/926

**Pull request merged**